File: 03-FEA-1D-IsoparametricElements/Matlab-Code/ReadInput.m
 function [N_NODE,N_ELEM,N_LOAD,N_PRE_DISP,...
       COORDS, ELEM_LOAD, ELEM_NODE,ELEM_STIFF, ...
       FORCE_NODE,FORCE_VAL,DISP_NODE,DISP_VAL, ...
       NNODE_ELE, DOF_NODE, EDOF, Ng, ShapeOrder] = ReadInput(filen)

% INPUT:
% filen: name of the file with input values
% nnode_ele: number of nodes per element

% OUTPUT: 
% N_NODE, the (integer) number of nodes
% N_ELEM, the (integer) number of elements
% N_LOAD, the (integer) number of nonzero nodal forces, i.e. Pf
% N_PRE_DISP, the (integer) number of nodes with prescribed displacement, i.e. Up
% ELEM_NODE(nnode_ele, N_ELEM), matrix that contains node numbers (integers) for each element
% ELEM_STIFF(N_ELEM), vector that contains stiffness value (real) for each element
% FORCE_NODE(N_LOAD), vector that contains the node numbers (integer) where forces are applied
% FORCE_VAL(N_LOAD), vector that contains the value of the forces (real) corresponding to the numbers in FORCE_NODE
% DISP_NODE(N_PRE_DISP), vector that contains the node numbers (integer) where boundary conditions are imposed
% DISP_VAL(N_PRE_DISP),  vector that contains the value of the forces (real) corresponding to the numbers in DISP_NODE

% Open input file
fid=fopen(filen,'r');

% read first line
A = fscanf(fid,'%i %i %i %i %i %i /n', 6);
N_NODE = A(1);
N_ELEM = A(2);    
N_LOAD = A(3); 
N_PRE_DISP = A(4);
Ng = A(5);
ShapeOrder = A(6);

if (ShapeOrder == 1) NNODE_ELE = 2; %Number of nodes per element
elseif (ShapeOrder == 2) NNODE_ELE = 3;
elseif (ShapeOrder == 3) NNODE_ELE = 4;
else disp('Only linear, quadratic and cubic functions are implemented. Exit');return;
end

if (Ng > 4) disp('Max Ng = 4. Exit');return;
end
DOF_NODE = 1; %Number of degrees of freedom per node
EDOF = NNODE_ELE*DOF_NODE; %number of degrees of freedom per element


% initialize vectors
COORDS = zeros(N_NODE,1);
ELEM_NODE = zeros(NNODE_ELE,N_ELEM); 
ELEM_STIFF = zeros(N_ELEM,1);
ELEM_LOAD = zeros(N_ELEM,1);
FORCE_NODE = zeros(N_LOAD,1); 
FORCE_VAL = zeros(N_LOAD,1);
DISP_NODE = zeros(N_PRE_DISP,1); 
DISP_VAL = zeros(N_PRE_DISP,1); 

for i=1:N_NODE
   line = fscanf(fid, '%f %f /n',1);
   COORDS(i,1) = line(1);
end


for i=1:N_ELEM
   
   for j=1:NNODE_ELE
        line = fscanf(fid, '%f',1);
        ELEM_NODE(j,i)=line(1);
   end
   
   line = fgetl(fid);
   linet = regexp(line, '[  |\t]', 'split');
   [a,b]=size(linet);

   j = 1;
   pos = 1;
   while j<=b
       
        string = linet{j};
        if isempty(string) 
            j=j+1;
            continue; 
        end
        if (pos==1)
            if strcmp('UMAT',string)
                ELEM_STIFF(i,1) = sqrt(-1); 
            else
                ELEM_STIFF(i,1) = str2double(string);
            end  
            pos = pos +1;
            j = j +1;
        elseif (pos==2)
            if strcmp('DLOAD',string)
                ELEM_LOAD(i,1) = sqrt(-1); 
            else
                ELEM_LOAD(i,1) = str2double(string);
            end  
            pos = pos +1;
            j = j + 1;
        end    
   end
   
end


% get prescribed force values and corresponding node num
for i=1:N_LOAD
   line = fscanf(fid, '%i %f /n',2);
   FORCE_NODE(i,1)=line(1);
   FORCE_VAL(i,1)=line(2);
end
% get prescribed displacement values and corresponding node num
for i=1:N_PRE_DISP
   line = fscanf(fid, '%i %f /n',2);  
   DISP_NODE(i,1)=line(1);
   DISP_VAL(i,1)=line(2);
end                         

fclose(fid);
end

File: 03-FEA-1D-IsoparametricElements/C-Code/ReadInput.cpp
#include "MainFile.h"

using namespace std;

void ReadInput(const std::string& filename, const int mdim, int& n_node, int& n_elem, int& n_load, int& n_pre_disp,
	       int& nnode_ele, int& dof_node, int& edof, int& Ng, int& ShapeOrder, arma::mat& coords, arma::imat& elem_node,
	       arma::colvec& elem_stiff, arma::colvec& elem_load, arma::ivec& UMATid, arma::ivec& DLOADid,
	       arma::icolvec& force_node, arma::colvec& force_val, arma::icolvec& disp_node, arma::colvec& disp_val)
{


  ifstream fout(filename.c_str());
  int      count = 0;

if (fout.is_open() ) {

  fout >> n_node >> n_elem >> n_load >> n_pre_disp >> Ng >> ShapeOrder;
  //cout<<n_node<<n_elem<<n_load<<n_pre_disp<<Ng<<ShapeOrder<<"\n";

  if (ShapeOrder == 1) nnode_ele = 2; //Number of nodes per element
  else if (ShapeOrder == 2) nnode_ele  = 3;
  else if (ShapeOrder == 3) nnode_ele  = 4;
  else {
    cout<<"Only linear, quadratic and cubic functions are implemented. Exit\n";
    exit(1);
  }

  if (Ng > 4) {
    cout<< "Input is Ng = "<<Ng<<". Max Ng = 4. Exit\n";
    exit(1);
  }
  dof_node  = 1; //Number of degrees of freedom per node
  edof = nnode_ele*dof_node;


  elem_node = arma::zeros<arma::imat>(nnode_ele,n_elem);
  elem_stiff = arma::zeros<arma::colvec>(n_elem);
  elem_load = arma::zeros<arma::colvec>(n_elem);
  UMATid = arma::zeros<arma::ivec>(n_elem);
  DLOADid = arma::zeros<arma::ivec>(n_elem);
  force_node  = arma::zeros<arma::icolvec>(n_load);
  force_val  = arma::zeros<arma::colvec>(n_load);
  disp_node  = arma::zeros<arma::icolvec>(n_pre_disp);
  disp_val  = arma::zeros<arma::colvec>(n_pre_disp);
  coords = arma::zeros<arma::mat>(n_node,mdim);



  for (int i=0;i<n_node;i++ ) {

    fout >> coords(i,0);

   }

  string e1,b1;
  for (int i=0;i<n_elem;i++ ) {
    for (int j=0;j<nnode_ele;j++) {
      fout >> elem_node(j,i);
    }
    fout >> e1  >> b1;
    if (e1 == "UMAT") {
      UMATid(i) = 1;
    }
    else {
      elem_stiff(i) = atof(e1.c_str());
    }
    if (b1 == "DLOAD") {
      DLOADid(i) = 1;
    }
    else {
      elem_load(i) = atof(b1.c_str());
    }
  }

  /*cout<<coords;
  cout<<elem_node;
  cout<<elem_stiff;
  cout<<UMATid;
  cout<<elem_load;
  cout<<DLOADid;
  cout<<Ng<<'\n';
  cout<<ShapeOrder;*/



  for (int i=0;i<n_load;i++ ) {

     fout >> force_node(i) >> force_val(i);

   }
  for (int i=0;i<n_pre_disp;i++ ) {

    fout >> disp_node(i) >> disp_val(i);

   }

  /*cout<<force_node;
  cout<<force_val;
  cout<<disp_node;
  cout<<disp_val;*/
}

 fout.close();

 cout<<"Completed Reading input.dat file\n";

}